File: runners/go/schema.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package main

import (
	"fmt"
	"regexp"
	"strconv"

	"github.com/apache/iceberg-go"
	"gopkg.in/yaml.v3"
)

// rowKeyName is the reserved synthetic identity column carried on every row so
// the orchestrator can compare row multisets across physical layouts. It is
// materialized as a real string column inside the table; the emitter relabels
// its field-id to the canonical 0 in output.
const rowKeyName = "__rowkey"

// rowKeyFieldID is the internal iceberg field-id assigned to __rowkey. User
// field-ids in the corpus are small (1..~1001 including partition field-ids);
// this id sits well above them but clear of iceberg's 2147483xxx reserved
// metadata-column range. The emitter maps this back to canonical field-id 0.
const rowKeyFieldID = 100000

var (
	decimalRe = regexp.MustCompile(`^decimal\(\s*(\d+)\s*,\s*(\d+)\s*\)$`)
	fixedRe   = regexp.MustCompile(`^fixed\[\s*(\d+)\s*\]$`)
)

// buildCanonIDs maps each top-level column name to the canonical output
// field-id used in the golden: __rowkey is 0, user columns keep the field-id
// the fixture declared. iceberg-go reassigns fresh field-ids at CreateTable
// (metadata.go AssignFreshSchemaIDs), so the runner must relabel by name back
// to these authored ids rather than emit the table's internal ids.
func buildCanonIDs(ls LSchema) map[string]int {
	m := map[string]int{rowKeyName: 0}
	for _, f := range ls.Fields {
		m[f.Name] = f.ID
	}
	return m
}

// buildSchema turns the L-log schema block into an iceberg.Schema, prepending
// the synthetic __rowkey string column. Returns the schema and the ordered
// list of user field names (schema order) for row materialization.
func buildSchema(ls LSchema) (*iceberg.Schema, []string, error) {
	fields := []iceberg.NestedField{{
		ID:       rowKeyFieldID,
		Name:     rowKeyName,
		Type:     iceberg.PrimitiveTypes.String,
		Required: false,
	}}
	names := make([]string, 0, len(ls.Fields))
	for _, f := range ls.Fields {
		nf, err := buildField(f)
		if err != nil {
			return nil, nil, err
		}
		fields = append(fields, nf)
		names = append(names, f.Name)
	}
	return iceberg.NewSchema(0, fields...), names, nil
}

// buildField resolves one L-log field (including initial/write defaults) into
// an iceberg.NestedField.
func buildField(f LField) (iceberg.NestedField, error) {
	typ, err := resolveType(&f.Type)
	if err != nil {
		return iceberg.NestedField{}, fmt.Errorf("field %q: %w", f.Name, err)
	}
	nf := iceberg.NestedField{
		ID:       f.ID,
		Name:     f.Name,
		Type:     typ,
		Required: f.Required,
		Doc:      f.Doc,
	}
	if f.InitialDefault != nil {
		lit, err := f.InitialDefault.toLiteral(typ)
		if err != nil {
			return iceberg.NestedField{}, fmt.Errorf("field %q initial-default: %w", f.Name, err)
		}
		nf.InitialDefault = lit.Any()
	}
	if f.WriteDefault != nil {
		lit, err := f.WriteDefault.toLiteral(typ)
		if err != nil {
			return iceberg.NestedField{}, fmt.Errorf("field %q write-default: %w", f.Name, err)
		}
		nf.WriteDefault = lit.Any()
	}
	return nf, nil
}

// resolveType maps an L-log type node (a primitive string like "long" /
// "decimal(9,2)" / "fixed[4]", or a nested struct/list/map object) to an
// iceberg.Type.
func resolveType(node *yaml.Node) (iceberg.Type, error) {
	switch node.Kind {
	case yaml.ScalarNode:
		return primitiveType(node.Value)
	case yaml.MappingNode:
		var probe struct {
			Type string `yaml:"type"`
		}
		if err := node.Decode(&probe); err != nil {
			return nil, err
		}
		switch probe.Type {
		case "struct":
			return resolveStruct(node)
		case "list":
			return resolveList(node)
		case "map":
			return resolveMap(node)
		default:
			return nil, fmt.Errorf("unknown nested type %q", probe.Type)
		}
	default:
		return nil, fmt.Errorf("unsupported type node kind %d", node.Kind)
	}
}

// primitiveType maps a spec primitive type name to its iceberg.Type.
func primitiveType(name string) (iceberg.Type, error) {
	switch name {
	case "boolean":
		return iceberg.PrimitiveTypes.Bool, nil
	case "int":
		return iceberg.PrimitiveTypes.Int32, nil
	case "long":
		return iceberg.PrimitiveTypes.Int64, nil
	case "float":
		return iceberg.PrimitiveTypes.Float32, nil
	case "double":
		return iceberg.PrimitiveTypes.Float64, nil
	case "date":
		return iceberg.PrimitiveTypes.Date, nil
	case "time":
		return iceberg.PrimitiveTypes.Time, nil
	case "timestamp":
		return iceberg.PrimitiveTypes.Timestamp, nil
	case "timestamptz":
		return iceberg.PrimitiveTypes.TimestampTz, nil
	case "timestamp_ns":
		return iceberg.PrimitiveTypes.TimestampNs, nil
	case "timestamptz_ns":
		return iceberg.PrimitiveTypes.TimestampTzNs, nil
	case "string":
		return iceberg.PrimitiveTypes.String, nil
	case "uuid":
		return iceberg.PrimitiveTypes.UUID, nil
	case "binary":
		return iceberg.PrimitiveTypes.Binary, nil
	}
	if m := decimalRe.FindStringSubmatch(name); m != nil {
		prec, _ := strconv.Atoi(m[1])
		scale, _ := strconv.Atoi(m[2])
		return iceberg.DecimalTypeOf(prec, scale), nil
	}
	if m := fixedRe.FindStringSubmatch(name); m != nil {
		n, _ := strconv.Atoi(m[1])
		return iceberg.FixedTypeOf(n), nil
	}
	return nil, fmt.Errorf("unsupported primitive type %q", name)
}

func resolveStruct(node *yaml.Node) (iceberg.Type, error) {
	var st struct {
		Fields []LField `yaml:"fields"`
	}
	if err := node.Decode(&st); err != nil {
		return nil, err
	}
	fields := make([]iceberg.NestedField, 0, len(st.Fields))
	for _, f := range st.Fields {
		nf, err := buildField(f)
		if err != nil {
			return nil, err
		}
		fields = append(fields, nf)
	}
	return &iceberg.StructType{FieldList: fields}, nil
}

func resolveList(node *yaml.Node) (iceberg.Type, error) {
	var lt struct {
		ElementID       int       `yaml:"element-id"`
		Element         yaml.Node `yaml:"element"`
		ElementRequired bool      `yaml:"element-required"`
	}
	if err := node.Decode(&lt); err != nil {
		return nil, err
	}
	elem, err := resolveType(&lt.Element)
	if err != nil {
		return nil, err
	}
	return &iceberg.ListType{
		ElementID:       lt.ElementID,
		Element:         elem,
		ElementRequired: lt.ElementRequired,
	}, nil
}

func resolveMap(node *yaml.Node) (iceberg.Type, error) {
	var mt struct {
		KeyID         int       `yaml:"key-id"`
		Key           yaml.Node `yaml:"key"`
		ValueID       int       `yaml:"value-id"`
		Value         yaml.Node `yaml:"value"`
		ValueRequired bool      `yaml:"value-required"`
	}
	if err := node.Decode(&mt); err != nil {
		return nil, err
	}
	keyType, err := resolveType(&mt.Key)
	if err != nil {
		return nil, err
	}
	valType, err := resolveType(&mt.Value)
	if err != nil {
		return nil, err
	}
	return &iceberg.MapType{
		KeyID:         mt.KeyID,
		KeyType:       keyType,
		ValueID:       mt.ValueID,
		ValueType:     valType,
		ValueRequired: mt.ValueRequired,
	}, nil
}

File: runners/rust/src/schema.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

//! Build an iceberg [`Schema`] from the op-log schema block, prepending the
//! synthetic `__rowkey` column, and map column names to the canonical output
//! field-ids used in the golden.

use std::collections::HashMap;
use std::sync::Arc;

use anyhow::{Result, anyhow, bail};
use iceberg::spec::{ListType, MapType, NestedField, PrimitiveType, Schema, StructType, Type};
use serde_yaml::Value as YamlValue;

use crate::llog::{LField, LSchema};
use crate::values::TypedValue;

/// Reserved synthetic identity column carried on every row so the orchestrator
/// can compare row multisets across physical layouts. Materialized as a real
/// string column; the emitter relabels its field-id to the canonical 0.
pub const ROW_KEY_NAME: &str = "__rowkey";

/// Internal iceberg field-id for `__rowkey`. Sits above the small corpus ids
/// but clear of iceberg's reserved 2147483xxx metadata-column range.
const ROW_KEY_FIELD_ID: i32 = 100_000;

/// Build the iceberg schema (with a leading `__rowkey` string column) from the
/// op-log schema block.
pub fn build_schema(ls: &LSchema) -> Result<Schema> {
    let mut fields: Vec<Arc<NestedField>> = Vec::with_capacity(ls.fields.len() + 1);
    fields.push(Arc::new(NestedField::optional(
        ROW_KEY_FIELD_ID,
        ROW_KEY_NAME,
        Type::Primitive(PrimitiveType::String),
    )));
    for f in &ls.fields {
        fields.push(Arc::new(build_field(f)?));
    }
    Schema::builder()
        .with_schema_id(0)
        .with_fields(fields)
        .build()
        .map_err(|e| anyhow!("schema build: {e}"))
}

/// Map each top-level column name to the canonical output field-id: `__rowkey`
/// is 0, user columns keep the id the fixture declared. iceberg-rust reassigns
/// fresh field-ids at create_table, so the emitter relabels by name back to
/// these authored ids.
pub fn build_canon_ids(ls: &LSchema) -> HashMap<String, i32> {
    let mut m = HashMap::new();
    m.insert(ROW_KEY_NAME.to_string(), 0);
    for f in &ls.fields {
        m.insert(f.name.clone(), f.id);
    }
    m
}

/// Derive canonical output field-ids from a loaded (read mode) table schema by
/// top-level column POSITION: __rowkey => 0, then user columns 1,2,... in schema
/// order. Read fixtures carry no authored ids and the impl's internal ids differ,
/// so position is the stable canonical labeling shared across implementations.
pub fn canon_ids_from_schema(schema: &Schema) -> HashMap<String, i32> {
    let mut m = HashMap::new();
    let mut next = 1;
    for f in schema.as_struct().fields() {
        if f.name == ROW_KEY_NAME {
            m.insert(f.name.clone(), 0);
        } else {
            m.insert(f.name.clone(), next);
            next += 1;
        }
    }
    m
}

fn build_field(f: &LField) -> Result<NestedField> {
    let ty = resolve_type(&f.field_type)?;
    let mut nf = if f.required {
        NestedField::required(f.id, &f.name, ty.clone())
    } else {
        NestedField::optional(f.id, &f.name, ty.clone())
    };
    if let Some(doc) = &f.doc {
        nf = nf.with_doc(doc);
    }
    if let Some(dv) = &f.initial_default {
        let tv = TypedValue::from_yaml(dv)?;
        nf = nf.with_initial_default(tv.to_literal(&ty)?);
    }
    if let Some(dv) = &f.write_default {
        let tv = TypedValue::from_yaml(dv)?;
        nf = nf.with_write_default(tv.to_literal(&ty)?);
    }
    Ok(nf)
}

/// Resolve an op-log type node (a primitive string like "long" /
/// "decimal(9,2)" / "fixed[4]", or a nested struct/list/map object) to a Type.
pub fn resolve_type(v: &YamlValue) -> Result<Type> {
    match v {
        YamlValue::String(s) => primitive_type(s),
        YamlValue::Mapping(_) => {
            let kind = v
                .get("type")
                .and_then(|t| t.as_str())
                .ok_or_else(|| anyhow!("nested type missing 'type'"))?;
            match kind {
                "struct" => resolve_struct(v),
                "list" => resolve_list(v),
                "map" => resolve_map(v),
                other => bail!("unknown nested type {other:?}"),
            }
        }
        other => bail!("unsupported type node {other:?}"),
    }
}

fn primitive_type(name: &str) -> Result<Type> {
    let p = match name {
        "boolean" => PrimitiveType::Boolean,
        "int" => PrimitiveType::Int,
        "long" => PrimitiveType::Long,
        "float" => PrimitiveType::Float,
        "double" => PrimitiveType::Double,
        "date" => PrimitiveType::Date,
        "time" => PrimitiveType::Time,
        "timestamp" => PrimitiveType::Timestamp,
        "timestamptz" => PrimitiveType::Timestamptz,
        "timestamp_ns" => PrimitiveType::TimestampNs,
        "timestamptz_ns" => PrimitiveType::TimestamptzNs,
        "string" => PrimitiveType::String,
        "uuid" => PrimitiveType::Uuid,
        "binary" => PrimitiveType::Binary,
        _ => {
            if let Some(rest) = name
                .strip_prefix("decimal(")
                .and_then(|s| s.strip_suffix(')'))
            {
                let (p, s) = rest
                    .split_once(',')
                    .ok_or_else(|| anyhow!("bad decimal type {name:?}"))?;
                return Ok(Type::Primitive(PrimitiveType::Decimal {
                    precision: p.trim().parse()?,
                    scale: s.trim().parse()?,
                }));
            }
            if let Some(n) = name
                .strip_prefix("fixed[")
                .and_then(|s| s.strip_suffix(']'))
            {
                return Ok(Type::Primitive(PrimitiveType::Fixed(n.trim().parse()?)));
            }
            bail!("unsupported primitive type {name:?}");
        }
    };
    Ok(Type::Primitive(p))
}

fn resolve_struct(v: &YamlValue) -> Result<Type> {
    let fields = v
        .get("fields")
        .and_then(|f| f.as_sequence())
        .ok_or_else(|| anyhow!("struct missing fields"))?;
    let mut out = Vec::with_capacity(fields.len());
    for fv in fields {
        let lf: LField = serde_yaml::from_value(fv.clone())?;
        out.push(Arc::new(build_field(&lf)?));
    }
    Ok(Type::Struct(StructType::new(out)))
}

fn resolve_list(v: &YamlValue) -> Result<Type> {
    let element_id = v
        .get("element-id")
        .and_then(|e| e.as_i64())
        .ok_or_else(|| anyhow!("list missing element-id"))? as i32;
    let element = resolve_type(
        v.get("element")
            .ok_or_else(|| anyhow!("list missing element"))?,
    )?;
    let required = v
        .get("element-required")
        .and_then(|b| b.as_bool())
        .unwrap_or(false);
    let field = if required {
        NestedField::required(element_id, "element", element)
    } else {
        NestedField::optional(element_id, "element", element)
    };
    Ok(Type::List(ListType::new(Arc::new(field))))
}

fn resolve_map(v: &YamlValue) -> Result<Type> {
    let key_id = v
        .get("key-id")
        .and_then(|e| e.as_i64())
        .ok_or_else(|| anyhow!("map missing key-id"))? as i32;
    let value_id = v
        .get("value-id")
        .and_then(|e| e.as_i64())
        .ok_or_else(|| anyhow!("map missing value-id"))? as i32;
    let key = resolve_type(v.get("key").ok_or_else(|| anyhow!("map missing key"))?)?;
    let value = resolve_type(v.get("value").ok_or_else(|| anyhow!("map missing value"))?)?;
    let required = v
        .get("value-required")
        .and_then(|b| b.as_bool())
        .unwrap_or(false);
    let key_field = Arc::new(NestedField::required(key_id, "key", key));
    let value_field = if required {
        NestedField::required(value_id, "value", value)
    } else {
        NestedField::optional(value_id, "value", value)
    };
    Ok(Type::Map(MapType::new(key_field, Arc::new(value_field))))
}
